benchmarks/c/miscellaneous/recursion.c

```diff
@@ -0,0 +1,70 @@
+#include <assert.h>
+#include <pthread.h>
+#include <stdatomic.h>
+int fib(int);
+int fibonacci(int n)
+{
+    return n == 0 ? 1 : n == 1 ? 1 : fib(n - 1) + fibonacci(n - 2);
+}
+int fib(int n)
+{
+    return fibonacci(n);
+}
+// known values: ack(0,0) == 1, ack(1,1) == 3, ack(2,2) == 7, ack(3,3) == 61
+int ack(int,int);
+int ackermann(int m, int n)
+{
+    return m == 0 ? n + 1 : n == 0 ? ack(m - 1, 1) : ack(m - 1, ack(m, n - 1));
+}
+int ack(int m, int n)
+{
+    return ackermann(m, n);
+}
+int fun(int n)
+{
+#if defined(ACKERMANN_LARGE) || defined(ACKERMANN_SMALL)
+    return ackermann(n, n);
+#else
+    return fibonacci(n);
+#endif
+}
+atomic_int x;
+void* worker(void* p)
+{
+    (void) p;
+    int n = atomic_load_explicit(&x, memory_order_relaxed);
+    int a = fun(n);
+    atomic_store_explicit(&x, a, memory_order_relaxed);
+    return NULL;
+}
+int main()
+{
+    pthread_t t0;
+#ifdef ACKERMANN_LARGE
+    atomic_init(&x, 1);
+#else
+#ifdef ACKERMANN_SMALL
+    atomic_init(&x, 0);
+#else//fib(fib(2)) == 2
+    atomic_init(&x, 2);
+#endif
+#endif
+    pthread_create(&t0, NULL, worker, NULL);
+    int n = atomic_load_explicit(&x, memory_order_relaxed);
+    int a = fun(n);
+    atomic_store_explicit(&x, a, memory_order_relaxed);
+    int result = atomic_load_explicit(&x, memory_order_relaxed);
+    pthread_join(t0, NULL);
+#ifdef ACKERMANN_LARGE
+    //ack(ack(1,1),ack(1,1)) needs a recursion bound of 64.
+    assert(result != 61);
+#else
+#ifdef ACKERMANN_SMALL
+    //ack(ack(0,0),ack(0,0) needs recursion bound of 4 and mem2reg after unrolling to be feasible.
+    assert(result != 3);
```

Comment:
Yes, doing the tests for PASS would be better.
Since inlining happens in one phase and in-place, recursive functions behave complicatedly:  Each recursive call copies the current state of the function into itself.  This means that the resulting depth in recursive functions grows exponentially with the provided bound.  This particular test has the following call tree.  For `fibonacci(2)`, there is apparently already enough program logic to avoid all bound events.
```
fibonacci (depth = 0)
  fib (depth = 0)
    fibonacci (depth = 1)
      fib (already inlined)
        fibonacci (depth = 2, BOUND)
      fibonacci (depth = 2, BOUND)
  fibonacci (depth = 1)
    fib (already inlined)
      fibonacci (already inlined)
        fib (already inlined)
          BOUND
        BOUND
    fibonacci (depth = 2, BOUND)
```

Reply:
Because of this exponential blowup I thought you introduced the cache originally so that you just inline the original state of the function repeatedly (this of course requires this pass to be a program pass rather than a function pass). I think we should avoid exponential blowup, but it is fine for now cause it's still better than not supporting recursion at all.

Reply:
The current version now does inlining based on copies of the original functions.  This way, each call has a predictable replacement and does not grow exponentially anymore.  Recursion depth is already counted per-function, so non-recursive programs will still be completely expanded.  I set the default recursion bound to 1, so each function may appear at most once in the call stack of the checked program fragment.  A recursion bound of 0 would mean that every call is replaced with a bound marker.